src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+        /// </summary>
+        virtual public IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer, Predicate<string> isEnabled)
+        {
+            // If we have been disposed, we silently ingore any subscriptions.  
+            if (m_disposed)
+                return new Subscription() { Owner = this };
+
+            Subscription newSubscription = new Subscription() { Observer = observer, IsEnabled = isEnabled, Owner = this, Next = m_subscriptions };
+            while (Interlocked.CompareExchange(ref m_subscriptions, newSubscription, newSubscription.Next) != newSubscription.Next)
+                newSubscription.Next = m_subscriptions;
+            return newSubscription;
+        }
+        public IDisposable Subscribe(IObserver<KeyValuePair<string, object>> observer)
+        {
+            return Subscribe(observer, null);
+        }
+
+        /// <summary>
+        /// Make a new TelemetryListener, it is a NotificationSource, which means the returned result can be used to 
+        /// log notifications, but it also has a Subscribe method so notifications can be forwarded
+        /// arbitrarily.  Thus its job is to forward things from the producer to all the listeners
+        /// (multi-casting).    Generally you should not be making your own TelemetryListener but use the
+        /// TelemetryListener.Default, so that notifications are as 'public' as possible.  
+        /// </summary>
+        public TelemetryListener(string name)
+        {
+            Name = name;
+            // To avoid allocating an explicit lock object I lock the Default TelemetryListener.   However there is a 
+            // chicken and egg problem because I need to call this code to initialize Default.   
+            var lockObj = DefaultListener;
+            if (lockObj == null)
+                lockObj = this;
+
+            // Insert myself into the list of all Listeners.   
+            lock (lockObj)
+            {
+                // Issue the callback for this new telemetry listener. 
+                var callback = s_allListenersCallback;
+                if (callback != null)
+                    callback(this);
+
+                // And add it to the list of all past listeners.  
+                m_next = s_allListeners;
+                s_allListeners = this;
+            }
+        }
+
+        /// <summary>
+        /// Clean up the NotificationListeners.   Notification listeners do NOT DIE ON THEIR OWN
+        /// because they are in a global list (for discoverability).  You must dispose them explicitly. 
+        /// Note that we do not do the Dispose(bool) pattern because we frankly don't want to support
+        /// subclasses that have non-managed state.   
+        /// </summary>
+        virtual public void Dispose()
+        {
+            // Remove myself from the list if all listeners.  
+            lock (DefaultListener)
+            {
+                m_disposed = true;
```

Comment:
You probably want to do ```if (_disposed) return;``` here in order to more efficiently support double-dispose and to avoid sending out duplicate OnCompleted messages below?